Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +125,141 @@
 	% This is the original order
-	vectorOldOrder = 1:nEncodedVolumes;
+	vectorReorderEncoded = 1:nEncodedVolumes;
 	
 	% Shape to a matrix with TEs first and all the rest later
-	vectorOldOrder = reshape(vectorOldOrder, xQ.NumberEchoTimes, nEncodedVolumes/xQ.NumberEchoTimes);
+	vectorReorderEncoded = reshape(vectorReorderEncoded, xQ.NumberEchoTimes, nEncodedVolumes/xQ.NumberEchoTimes);
 	
 	% Flip the two dimensions and make a row vector again
-	vectorOldOrder = reshape(vectorOldOrder', 1, nEncodedVolumes);
+	vectorReorderEncoded = reshape(vectorReorderEncoded', 1, nEncodedVolumes);
 
 	% Reorder the data
-	imEncoded = imEncoded(:,:,:,vectorOldOrder);
+	imEncodedReordered = imEncoded(:,:,:,vectorReorderEncoded);
 end
 
 % Sometimes HAD8 is PLD1(rep1), PLD1(rep2), PLD2(rep1), PLD2(rep2). We
 % want: PLD1(rep1), PLD2(rep1)... PLD1(rep2),PLD2(rep2)
-% UniquePLDs = unique(xQ.Initial_PLD);
-% 
-% if (length(UniquePLDs) == nEncodedVolumes) && (UniquePLDs(2)-UniquePLDs(1))< 250
-%     This is the original order
-% 	vectorOldOrder = 1:nEncodedVolumes;
-%     Shape
-%     vectorOrder = [ vectorOldOrder(1:2:end) vectorOldOrder(2:2:end)];
-%     Reorder the data
-% 	imEncoded = imEncoded(:,:,:,vectorOrder);
-% end
-
-%% 3. Decode the Hadamard data
-imDecoded = zeros(size(imEncoded,1), size(imEncoded,2), size(imEncoded,3), nDecodedVolumes * nRepetitions);    
-idx=0;
-for iRepetition = 1:nRepetitions
-    for iTE = 1:xQ.NumberEchoTimes
-        for iTI = 1:nDecodedTI
-            
-            indexPositive = find(xQ.TimeEncodedMatrix(iTI,:)==1);
-            indexNegative = find(xQ.TimeEncodedMatrix(iTI,:)==-1);
-            imDecoded(:,:,:,((iTE-1)*nDecodedTI+iTI)+(iRepetition-1)*nDecodedVolumes) = sum(imEncoded(:,:,:,(indexPositive+idx)),4) - sum(imEncoded(:,:,:,(indexNegative+idx)),4);
-            
-        end
-        idx = idx+xQ.TimeEncodedMatrixSize;
-    end
-end
 
+%% 5. Decode the Hadamard data
+% We assume that the entire encoded volume is reorder into several repetitions of basic Hadamard blocks (repetitions or TE are counted the same)
+decodedPWI4D = zeros(size(imEncodedReordered,1), size(imEncodedReordered,2), size(imEncodedReordered,3), nEncodedVolumes / xQ.TimeEncodedMatrixSize * (xQ.TimeEncodedMatrixSize-1)); 
+decodedBlock = zeros(size(imEncodedReordered,1), size(imEncodedReordered,2), size(imEncodedReordered,3), xQ.TimeEncodedMatrixSize-1); % Temporary single decoded Hadamard block
 
-%% 4. Reorder multi-TE back to the initial order of PLD/TE
-%
-% For model fitting, we want the PLDs-first-TEs-second order (just like the
-% beginning) so we need to reorder it again
+% Go through all repetitions of the Hadamard blocks
+for iRepetition = 1:(nEncodedVolumes / xQ.TimeEncodedMatrixSize)
+	offsetEncoded = xQ.TimeEncodedMatrixSize * (iRepetition-1); % Index offset for the given repetition to search in the encoded data
+	encodedBlock = imEncoded(:,:,:,(1:xQ.TimeEncodedMatrixSize) + offsetEncoded); % Obtain the single encoded block
```

Comment:
Each repetition has multiple blocks right, so shouldn't decodedBlock then be newly initialized per Repetition?

Reply:
2. I know that this is what it should be. But for debugging purposes, I would do the initialization in the loop instead of outside the loop. Just easier to read/debug right? (in fact, the whole initialization is extra code for no reason, it will affect 1 millisecond speed unless you're running ExploreASL on a raspberry ;)

Reply:
Renamed to blocks

Moved decodedBlocks init inside the cycle for clarity.

Done in 0a25d254